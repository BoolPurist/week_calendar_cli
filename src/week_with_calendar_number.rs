use chrono::{Datelike, Local, NaiveDate, Weekday};
use date_validation_types::{ValidatedDate, ValidatedDay, ValidatedMonth, ValidatedYear};
use derive_getters::Getters;
use prettytable::{row, Row};

#[derive(Getters, Debug)]
pub struct WeekWithCalendarNumber {
    week: u32,
    monday: NaiveDate,
    sunday: NaiveDate,
}

impl WeekWithCalendarNumber {
    pub fn new(date: impl Into<NaiveDate>) -> Self {
        let date: NaiveDate = date.into();
        let first_day_of_year = NaiveDate::from_ymd_opt(date.year(), 1, 1).unwrap();
        let week = date.week(Weekday::Mon);
        let monday = week.first_day();
        let sunday = week.last_day();
        let week = monday - first_day_of_year;
        let week = week.num_weeks() as u32 + 1;
        Self {
            week,
            monday,
            sunday,
        }
    }

    pub fn today() -> Self {
        let today: NaiveDate = Local::now().naive_local().date();
        today.into()
    }

    pub fn all_new_in_month(date: impl Into<NaiveDate>) -> Vec<Self> {
        let date: NaiveDate = date.into();
        dbg!(date);
        let last_date_of_month = Self::get_last_day_current_month(date);
        let month = date.month();
        let first_week = NaiveDate::from_ymd_opt(date.year(), month, 1).unwrap();
        let mut current_week = first_week.week(Weekday::Mon).first_day();
        let mut weeks: Vec<Self> = Vec::with_capacity(6);

        while current_week <= last_date_of_month {
            weeks.push(Self::new(current_week));
            current_week = current_week.week(Weekday::Mon).last_day() + chrono::Duration::days(1);
        }

        weeks
    }

    fn get_last_day_current_month(date: NaiveDate) -> NaiveDate {
        let (next_month, next_year) = {
            let mut next_year = date.year();
            let next_month = if date.month() == 12 {
                next_year += 1;
                1
            } else {
                date.month() + 1
            };

            (next_month, next_year)
        };

        NaiveDate::from_ymd_opt(next_year, next_month, 1).unwrap() - chrono::Duration::days(1)
    }
}

impl From<WeekWithCalendarNumber> for Row {
    fn from(value: WeekWithCalendarNumber) -> Self {
        row![value.week, value.monday, value.sunday]
    }
}
impl From<NaiveDate> for WeekWithCalendarNumber {
    fn from(value: NaiveDate) -> Self {
        let year = value.year() as u32;
        let year: ValidatedYear = year.try_into().unwrap();
        let month: ValidatedMonth = value.month().try_into().unwrap();
        let day: ValidatedDay = value.day().try_into().unwrap();
        Self::new(ValidatedDate::new(year, month, day).unwrap())
    }
}

#[cfg(test)]
mod testing {
    use super::*;
    #[test]
    fn new_with_calendar_week_and_monday_and_sunday() {
        let actual = WeekWithCalendarNumber::new(ValidatedDate::from_ymd(2023, 6, 30).unwrap());

        assert_eq!(26, actual.week);
        assert_eq!(NaiveDate::from_ymd_opt(2023, 6, 26).unwrap(), actual.monday);
        assert_eq!(NaiveDate::from_ymd_opt(2023, 7, 2).unwrap(), actual.sunday);
    }
    #[test]
    fn new_with_calendar_single_week_1_1_2020() {
        let actual = WeekWithCalendarNumber::new(ValidatedDate::from_ymd(2020, 1, 1).unwrap());

        assert_eq!(1, actual.week);
        assert_eq!(
            NaiveDate::from_ymd_opt(2019, 12, 30).unwrap(),
            actual.monday
        );
        assert_eq!(NaiveDate::from_ymd_opt(2020, 1, 5).unwrap(), actual.sunday);
    }
    #[test]
    fn new_within_monts() {
        let actual =
            WeekWithCalendarNumber::all_new_in_month(ValidatedDate::from_ymd(2023, 6, 30).unwrap());
        insta::assert_debug_snapshot!(actual);
    }
    #[test]
    fn new_within_months_from_earlier_year() {
        let actual =
            WeekWithCalendarNumber::all_new_in_month(ValidatedDate::from_ymd(2023, 1, 1).unwrap());
        insta::assert_debug_snapshot!(actual);
    }
    #[test]
    fn new_within_months_at_end_of_year_2023() {
        let actual = WeekWithCalendarNumber::all_new_in_month(
            NaiveDate::from_ymd_opt(2023, 12, 29).unwrap(),
        );
        insta::assert_debug_snapshot!(actual);
    }
    #[test]
    fn new_within_months_at_end_of_year_2022() {
        let actual = WeekWithCalendarNumber::all_new_in_month(
            NaiveDate::from_ymd_opt(2022, 12, 10).unwrap(),
        );
        insta::assert_debug_snapshot!(actual);
    }
}
