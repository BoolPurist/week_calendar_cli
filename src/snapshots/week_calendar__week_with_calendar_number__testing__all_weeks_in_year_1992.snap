---
source: src/week_with_calendar_number.rs
expression: result
---
[
    WeekWithCalendarNumber {
        week: 52,
        monday: 1991-12-30,
        sunday: 1992-01-05,
    },
    WeekWithCalendarNumber {
        week: 1,
        monday: 1992-01-06,
        sunday: 1992-01-12,
    },
    WeekWithCalendarNumber {
        week: 2,
        monday: 1992-01-13,
        sunday: 1992-01-19,
    },
    WeekWithCalendarNumber {
        week: 3,
        monday: 1992-01-20,
        sunday: 1992-01-26,
    },
    WeekWithCalendarNumber {
        week: 4,
        monday: 1992-01-27,
        sunday: 1992-02-02,
    },
    WeekWithCalendarNumber {
        week: 5,
        monday: 1992-02-03,
        sunday: 1992-02-09,
    },
    WeekWithCalendarNumber {
        week: 6,
        monday: 1992-02-10,
        sunday: 1992-02-16,
    },
    WeekWithCalendarNumber {
        week: 7,
        monday: 1992-02-17,
        sunday: 1992-02-23,
    },
    WeekWithCalendarNumber {
        week: 8,
        monday: 1992-02-24,
        sunday: 1992-03-01,
    },
    WeekWithCalendarNumber {
        week: 9,
        monday: 1992-03-02,
        sunday: 1992-03-08,
    },
    WeekWithCalendarNumber {
        week: 10,
        monday: 1992-03-09,
        sunday: 1992-03-15,
    },
    WeekWithCalendarNumber {
        week: 11,
        monday: 1992-03-16,
        sunday: 1992-03-22,
    },
    WeekWithCalendarNumber {
        week: 12,
        monday: 1992-03-23,
        sunday: 1992-03-29,
    },
    WeekWithCalendarNumber {
        week: 13,
        monday: 1992-03-30,
        sunday: 1992-04-05,
    },
    WeekWithCalendarNumber {
        week: 14,
        monday: 1992-04-06,
        sunday: 1992-04-12,
    },
    WeekWithCalendarNumber {
        week: 15,
        monday: 1992-04-13,
        sunday: 1992-04-19,
    },
    WeekWithCalendarNumber {
        week: 16,
        monday: 1992-04-20,
        sunday: 1992-04-26,
    },
    WeekWithCalendarNumber {
        week: 17,
        monday: 1992-04-27,
        sunday: 1992-05-03,
    },
    WeekWithCalendarNumber {
        week: 18,
        monday: 1992-05-04,
        sunday: 1992-05-10,
    },
    WeekWithCalendarNumber {
        week: 19,
        monday: 1992-05-11,
        sunday: 1992-05-17,
    },
    WeekWithCalendarNumber {
        week: 20,
        monday: 1992-05-18,
        sunday: 1992-05-24,
    },
    WeekWithCalendarNumber {
        week: 21,
        monday: 1992-05-25,
        sunday: 1992-05-31,
    },
    WeekWithCalendarNumber {
        week: 22,
        monday: 1992-06-01,
        sunday: 1992-06-07,
    },
    WeekWithCalendarNumber {
        week: 23,
        monday: 1992-06-08,
        sunday: 1992-06-14,
    },
    WeekWithCalendarNumber {
        week: 24,
        monday: 1992-06-15,
        sunday: 1992-06-21,
    },
    WeekWithCalendarNumber {
        week: 25,
        monday: 1992-06-22,
        sunday: 1992-06-28,
    },
    WeekWithCalendarNumber {
        week: 26,
        monday: 1992-06-29,
        sunday: 1992-07-05,
    },
    WeekWithCalendarNumber {
        week: 27,
        monday: 1992-07-06,
        sunday: 1992-07-12,
    },
    WeekWithCalendarNumber {
        week: 28,
        monday: 1992-07-13,
        sunday: 1992-07-19,
    },
    WeekWithCalendarNumber {
        week: 29,
        monday: 1992-07-20,
        sunday: 1992-07-26,
    },
    WeekWithCalendarNumber {
        week: 30,
        monday: 1992-07-27,
        sunday: 1992-08-02,
    },
    WeekWithCalendarNumber {
        week: 31,
        monday: 1992-08-03,
        sunday: 1992-08-09,
    },
    WeekWithCalendarNumber {
        week: 32,
        monday: 1992-08-10,
        sunday: 1992-08-16,
    },
    WeekWithCalendarNumber {
        week: 33,
        monday: 1992-08-17,
        sunday: 1992-08-23,
    },
    WeekWithCalendarNumber {
        week: 34,
        monday: 1992-08-24,
        sunday: 1992-08-30,
    },
    WeekWithCalendarNumber {
        week: 35,
        monday: 1992-08-31,
        sunday: 1992-09-06,
    },
    WeekWithCalendarNumber {
        week: 36,
        monday: 1992-09-07,
        sunday: 1992-09-13,
    },
    WeekWithCalendarNumber {
        week: 37,
        monday: 1992-09-14,
        sunday: 1992-09-20,
    },
    WeekWithCalendarNumber {
        week: 38,
        monday: 1992-09-21,
        sunday: 1992-09-27,
    },
    WeekWithCalendarNumber {
        week: 39,
        monday: 1992-09-28,
        sunday: 1992-10-04,
    },
    WeekWithCalendarNumber {
        week: 40,
        monday: 1992-10-05,
        sunday: 1992-10-11,
    },
    WeekWithCalendarNumber {
        week: 41,
        monday: 1992-10-12,
        sunday: 1992-10-18,
    },
    WeekWithCalendarNumber {
        week: 42,
        monday: 1992-10-19,
        sunday: 1992-10-25,
    },
    WeekWithCalendarNumber {
        week: 43,
        monday: 1992-10-26,
        sunday: 1992-11-01,
    },
    WeekWithCalendarNumber {
        week: 44,
        monday: 1992-11-02,
        sunday: 1992-11-08,
    },
    WeekWithCalendarNumber {
        week: 45,
        monday: 1992-11-09,
        sunday: 1992-11-15,
    },
    WeekWithCalendarNumber {
        week: 46,
        monday: 1992-11-16,
        sunday: 1992-11-22,
    },
    WeekWithCalendarNumber {
        week: 47,
        monday: 1992-11-23,
        sunday: 1992-11-29,
    },
    WeekWithCalendarNumber {
        week: 48,
        monday: 1992-11-30,
        sunday: 1992-12-06,
    },
    WeekWithCalendarNumber {
        week: 49,
        monday: 1992-12-07,
        sunday: 1992-12-13,
    },
    WeekWithCalendarNumber {
        week: 50,
        monday: 1992-12-14,
        sunday: 1992-12-20,
    },
    WeekWithCalendarNumber {
        week: 51,
        monday: 1992-12-21,
        sunday: 1992-12-27,
    },
    WeekWithCalendarNumber {
        week: 52,
        monday: 1992-12-28,
        sunday: 1993-01-03,
    },
]
